### ISSUE TITLE
django-allauth compatibility

### ISSUE THREAD

**Issue by grucha** [jazzband/django-axes#289]
For me `django-axes` didn't work with `django-allauth` package by default.

## Problem 1:

After some digging in code I've made my own `apps.py` that monkey patches allauth's `LoginView` the same way as axes does it to django's `LoginView`:
```
class AppConfig(apps.AppConfig):
    name = 'my_app_name'

    def ready(self):
        from allauth.account.views import LoginView  # <-- Just this line is changed
        from django.utils.decorators import method_decorator

        from axes import signals    # we must load signals
        from axes.decorators import axes_dispatch
        from axes.decorators import axes_form_invalid

        LoginView.dispatch = method_decorator(axes_dispatch)(LoginView.dispatch)
        LoginView.form_invalid = method_decorator(axes_form_invalid)(LoginView.form_invalid)
```

It may be useful to either point that out in readme or to modify `apps.py` to try to detect allauth and patch it too.

## Problem 2:
axes rely on having login id stored under `AXES_USERNAME_FORM_FIELD` key both in request.POST and in `credentials` arg of `user_login_failed` signal. This is not the case with allauth. They allways use `login` key in post POST data but it becomes `username` key in `credentials` dict in signal handler.

When I put `AXES_USERNAME_FORM_FIELD='login'` in use with `AXES_ONLY_USER_FAILURES=True` it causes empty username go to cache key and therefore locking everybody out when anyone is locked. They all get the same cache key then.

I'm using:
django-axes==4.0.1
django-allauth==0.34.0 (note: 0.31.0 didn't work with my fix either, had to upgrade)
django==1.11

**Comment by aleksihakli** [comment]
If you are using a third party or custom authentication package then you need to make sure the login views and other mechanisms they are using are either compatible with the way `django-axes` does things or write your own glue logic between the packages, this is case with [other packages as well](https://github.com/jazzband/django-axes/issues/287#issuecomment-355815146).

Thanks for reporting though, hope this makes it easier for somebody to solve the same thing again! If you want you can contribute to the documentation in the `docs` folder and add instructions for connecting custom login backends :)

**Comment by grucha** [comment]
For now I'll only paste my solution to **Problem 2** mentioned above:

In settings: `AXES_USERNAME_FORM_FIELD = 'login'`

In `my_app.forms.py`:
```
from allauth.account.forms import LoginForm

class AllauthCompatLoginForm(LoginForm):
    def user_credentials(self):
        """
        For `axes` and `allauth` compatibility we simply add "login" key to the creentials
        dict so it is consistent with POST's "login" data key. 
        """
        credentials = super(AllauthCompatLoginForm, self).user_credentials()
        credentials['login'] = credentials.get('email') or credentials.get('username')
        return credentials
```

And then use this custom form in `urls.py`:
```
from allauth.account import views as allauth_views
from my_app.forms import AllauthCompatLoginForm

urlpatterns = [
    ...
    url(r'^accounts/login/$', # Override allauth's default view with our compatibility mod
        allauth_views.LoginView.as_view(form_class=AllauthCompatLoginForm),
        name="account_login"),
    url(r'^accounts/', include('allauth.urls')),
]
```

Also, **Problem 1** might be solved in more elegant way by wrapping the view in `urls.py` as you mentioned in issue #287 

I hope this helps somebody.

**Comment by camilonova** [comment]
Don't you think it is a great idea to have this as a page at the documentation?

**Comment by aleksihakli** [comment]
@grucha could you please add a PR for adding this to the documentation in the `docs` folder so we can update RTD page? That way everyone browsing the package docs could benefit from your findings.

**Comment by pjmcdermott** [comment]
I have just tried the patches above. When I try a login with an **incorrect** password, I get a `KeyError: 'username'`. In `axes/signals.py` there is:

```
	@receiver(user_login_failed)
  	def log_user_login_failed(sender, credentials, request, **kwargs):
  	    """ Create an AccessAttempt record if the login wasn't successful
  	    """
  	    ip_address = get_ip(request)
  	    username = credentials['username']
            ....
```
which is causing the error.

I have `django-allauth` set to use the email field as the username using `ACCOUNT_AUTHENTICATION_METHOD = "email"`, so in the case of an incorrect password `django-axes` is trying to log a message with the username and failing because it is not in the credentials.

So I don't think the above suggestions solve the whole problem.

Ntwithstanding @aleksihakli's comments above regarding the need for glue logic, is there not a possible solution whereby `django-axes` has a similar setting to `django-allauth`, e.g. `AXES_ACCOUNT_AUTHENTICATION_METHOD`, which can be set to either 'username' or 'email'? Or possibly `django-axes` simply honours `django-allauth`'s `ACCOUNT_AUTHENTICATION_METHOD` setting, when set.

**Comment by grucha** [comment]
`axes` should honour it's own `AXES_USERNAME_FORM_FIELD` setting here. 
And it does for me - I'm using `django-axes==4.0.1` and the `axes/signals.py:42` line reads:

```username = credentials[settings.AXES_USERNAME_FORM_FIELD]```

**Comment by pjmcdermott** [comment]
@grucha Hmm...I seem to be on 3.0.3 for some reason. Thank for pointing this out.

**Comment by pjmcdermott** [comment]
I can confirm this `django-allauth` compatability tip is working for me **once on** `django-axes==4.0.1`.

**Comment by aleksihakli** [comment]
Fixed in #294.
